qiime2/core/type/meta.py

```diff
@@ -0,0 +1,286 @@
+        def select(x, swap=None):
+            if swap is not None:
+                return x.duplicate(predicate=swap)
+            return x.full_predicate
+
+        yield select
+
+    for idx, field in enumerate(expr.fields):
+        for sel in select_variables(field):
+            # Without this closure, the idx in select will be the last
+            # value of the enumerate, same for sel
+            # (Same problem as JS with callbacks inside a loop)
+            def closure(idx, sel):
+                def select(x, swap=None):
+                    if swap is not None:
+                        new_fields = list(x.fields)
+                        new_fields[idx] = sel(x.fields[idx], swap)
+                        return x.duplicate(fields=tuple(new_fields))
+                    return sel(x.fields[idx])
+                return select
+            yield closure(idx, sel)
+
+
+def match(provided, inputs, outputs):
+    provided_binding = {}
+    error_map = {}
+    for key, expr in inputs.items():
+        for selector in select_variables(expr):
+            var = selector(expr)
+            provided_fragment = selector(provided[key])
+            try:
+                current_binding = provided_binding[var]
+            except KeyError:
+                provided_binding[var] = provided_fragment
+                error_map[var] = provided[key]
+            else:
+                if not var.uniq_upto_sub(current_binding, provided_fragment):
+                    raise ValueError("Received %r and %r, but expected %r"
+                                     " and %r to match (or to select the same"
+                                     " output)."
+                                     % (error_map[var], provided[key],
+                                        current_binding, provided_fragment))
+
+    # provided_binding now maps TypeVarExp instances to a TypeExp instance
+    # which is the relevent fragment from the provided input types
+
+    grouped_maps = {}
+    for item in provided_binding.items():
+        var = item[0]
+        if var.mapping not in grouped_maps:
+            grouped_maps[var.mapping] = [item]
+        else:
+            grouped_maps[var.mapping].append(item)
+
+    # grouped_maps now maps a TypeMap instance to tuples of
+    # (TypeVarExp, TypeExp) which are the items of provided_binding
+    # i.e. all of the bindings are now grouped under their shared type maps
+
+    output_fragments = {}
+    for mapping, group in grouped_maps.items():
+        if len(group) != mapping.input_width():
+            raise ValueError("Missing input variables")
+
+        inputs = [x[1] for x in sorted(group, key=lambda x: x[0].index)]
+        solved = mapping.solve(*inputs)
+        if solved is None:
+            provided = tuple(error_map[x[0]]
+                             for x in sorted(group, key=lambda x: x[0].index))
+            raise ValueError("No solution for inputs: %r" % (provided,))
```

Comment:
I suspect this error message might be confusing to the poor user that sees it, maybe we should enumerate all the possible inputs in the message?